In [3]:
"""Сравнение DistilBERT vs BERT-base на одних и тех же split'ах."""
import time

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import torch
from datasets import Dataset
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, f1_score
from transformers import (
    AutoModelForSequenceClassification,
    AutoTokenizer,
    Trainer,
    TrainingArguments,
)

MODELS = [
    "distilbert-base-uncased",
    "bert-base-uncased",
]

MAX_LENGTH = 256
EPOCHS = 3
BATCH_SIZE = 8  # одинаково для обеих моделей
LEARNING_RATE = 2e-5

train_df = pd.read_csv("train.csv")
val_df = pd.read_csv("val.csv")
test_df = pd.read_csv("test.csv")

labels = sorted(train_df["label"].unique())
label2id = {label: i for i, label in enumerate(labels)}
id2label = {i: label for label, i in label2id.items()}


def to_dataset(df):
    return Dataset.from_dict({
        "text": df["text"].tolist(),
        "label": [label2id[x] for x in df["label"]],
    })


def compute_metrics(eval_pred):
    logits, y_true = eval_pred
    y_pred = np.argmax(logits, axis=1)
    return {
        "accuracy": accuracy_score(y_true, y_pred),
        "f1_macro": f1_score(y_true, y_pred, average="macro"),
    }


def train_and_evaluate(model_name: str) -> dict:
    print(f"\n{'=' * 60}")
    print(f"Обучение: {model_name}")
    print("=" * 60)

    train_ds = to_dataset(train_df)
    val_ds = to_dataset(val_df)
    test_ds = to_dataset(test_df)

    tokenizer = AutoTokenizer.from_pretrained(model_name)

    def tokenize(batch):
        return tokenizer(
            batch["text"],
            truncation=True,
            padding="max_length",
            max_length=MAX_LENGTH,
        )

    train_ds = train_ds.map(tokenize, batched=True)
    val_ds = val_ds.map(tokenize, batched=True)
    test_ds = test_ds.map(tokenize, batched=True)

    model = AutoModelForSequenceClassification.from_pretrained(
        model_name,
        num_labels=len(labels),
        id2label=id2label,
        label2id=label2id,
    )

    short_name = model_name.split("/")[-1]
    training_args = TrainingArguments(
        output_dir=f"./checkpoints_{short_name}",
        num_train_epochs=EPOCHS,
        per_device_train_batch_size=BATCH_SIZE,
        per_device_eval_batch_size=BATCH_SIZE,
        eval_strategy="epoch",
        save_strategy="epoch",
        load_best_model_at_end=True,
        metric_for_best_model="accuracy",
        learning_rate=LEARNING_RATE,
        logging_steps=50,
        report_to="none",
    )

    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_ds,
        eval_dataset=val_ds,
        compute_metrics=compute_metrics,
    )

    start = time.time()
    trainer.train()
    train_time_min = (time.time() - start) / 60

    preds = trainer.predict(test_ds)
    y_pred = np.argmax(preds.predictions, axis=1)
    y_true = test_df["label"].map(label2id).tolist()

    accuracy = accuracy_score(y_true, y_pred)
    f1_macro = f1_score(y_true, y_pred, average="macro")

    print(f"\n=== {model_name} ===")
    print("Accuracy:", round(accuracy, 3))
    print("F1 macro:", round(f1_macro, 3))
    print(f"Время обучения: {train_time_min:.1f} мин")
    print(classification_report(y_true, y_pred, target_names=labels))

    # Матрица
    cm = confusion_matrix(y_true, y_pred)
    plt.figure(figsize=(8, 6))
    sns.heatmap(cm, annot=True, fmt="d", xticklabels=labels, yticklabels=labels)
    plt.xlabel("Predicted")
    plt.ylabel("True")
    plt.title(f"Confusion matrix — {short_name}")
    plt.tight_layout()
    cm_path = f"confusion_matrix_{short_name}.png"
    plt.savefig(cm_path, dpi=150)
    plt.close()
    print(f"Сохранено: {cm_path}")

    # Освобождаем GPU перед следующей моделью
    del model, trainer, tokenizer
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

    return {
        "model": short_name,
        "accuracy": round(accuracy, 3),
        "f1_macro": round(f1_macro, 3),
        "train_time_min": round(train_time_min, 1),
        "epochs": EPOCHS,
        "batch_size": BATCH_SIZE,
    }


results = [train_and_evaluate(name) for name in MODELS]

comparison = pd.DataFrame(results)
print("\n" + "=" * 60)
print("ИТОГОВОЕ СРАВНЕНИЕ")
print("=" * 60)
print(comparison.to_string(index=False))

comparison.to_csv("model_comparison.csv", index=False)
print("\nСохранено: model_comparison.csv")



Обучение: distilbert-base-uncased


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:124: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Map:   0%|          | 0/2400 [00:00<?, ? examples/s]

Map:   0%|          | 0/300 [00:00<?, ? examples/s]

Map:   0%|          | 0/300 [00:00<?, ? examples/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/100 [00:00<?, ?it/s]

[transformers] DistilBertForSequenceClassification LOAD REPORT from: distilbert-base-uncased
Key                     | Status     | 
------------------------+------------+-
vocab_layer_norm.bias   | UNEXPECTED | 
vocab_transform.weight  | UNEXPECTED | 
vocab_projector.bias    | UNEXPECTED | 
vocab_layer_norm.weight | UNEXPECTED | 
vocab_transform.bias    | UNEXPECTED | 
pre_classifier.bias     | MISSING    | 
classifier.bias         | MISSING    | 
classifier.weight       | MISSING    | 
pre_classifier.weight   | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Epoch,Training Loss,Validation Loss,Accuracy,F1 Macro
1,0.670642,0.567600,0.790000,0.787718
2,0.375188,0.526493,0.826667,0.828166
3,0.202478,0.524705,0.833333,0.834589


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]


=== distilbert-base-uncased ===
Accuracy: 0.863
F1 macro: 0.863
Время обучения: 3.5 мин
                precision    recall  f1-score   support

       Fantasy       0.80      0.88      0.84        60
       History       0.88      0.88      0.88        60
    Nonfiction       0.89      0.85      0.87        60
       Romance       0.86      0.82      0.84        60
Sequential Art       0.88      0.88      0.88        60

      accuracy                           0.86       300
     macro avg       0.86      0.86      0.86       300
  weighted avg       0.86      0.86      0.86       300

Сохранено: confusion_matrix_distilbert-base-uncased.png

Обучение: bert-base-uncased


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Map:   0%|          | 0/2400 [00:00<?, ? examples/s]

Map:   0%|          | 0/300 [00:00<?, ? examples/s]

Map:   0%|          | 0/300 [00:00<?, ? examples/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

[transformers] BertForSequenceClassification LOAD REPORT from: bert-base-uncased
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.transform.dense.weight     | UNEXPECTED | 
cls.predictions.transform.dense.bias       | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
cls.predictions.bias                       | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
classifier.bias                            | MISSING    | 
classifier.weight                          | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Epoch,Training Loss,Validation Loss,Accuracy,F1 Macro
1,0.639156,0.537515,0.810000,0.808947
2,0.364811,0.546293,0.820000,0.820662
3,0.174278,0.599811,0.820000,0.820984


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

[transformers] There were missing keys in the checkpoint model loaded: ['bert.embeddings.LayerNorm.weight', 'bert.embeddings.LayerNorm.bias', 'bert.encoder.layer.0.attention.output.LayerNorm.weight', 'bert.encoder.layer.0.attention.output.LayerNorm.bias', 'bert.encoder.layer.0.output.LayerNorm.weight', 'bert.encoder.layer.0.output.LayerNorm.bias', 'bert.encoder.layer.1.attention.output.LayerNorm.weight', 'bert.encoder.layer.1.attention.output.LayerNorm.bias', 'bert.encoder.layer.1.output.LayerNorm.weight', 'bert.encoder.layer.1.output.LayerNorm.bias', 'bert.encoder.layer.2.attention.output.LayerNorm.weight', 'bert.encoder.layer.2.attention.output.LayerNorm.bias', 'bert.encoder.layer.2.output.LayerNorm.weight', 'bert.encoder.layer.2.output.LayerNorm.bias', 'bert.encoder.layer.3.attention.output.LayerNorm.weight', 'bert.encoder.layer.3.attention.output.LayerNorm.bias', 'bert.encoder.layer.3.output.LayerNorm.weight', 'bert.encoder.layer.3.output.LayerNorm.bias', 'bert.encoder.layer.4.atte


=== bert-base-uncased ===
Accuracy: 0.847
F1 macro: 0.849
Время обучения: 6.6 мин
                precision    recall  f1-score   support

       Fantasy       0.71      0.82      0.76        60
       History       0.94      0.80      0.86        60
    Nonfiction       0.81      0.93      0.87        60
       Romance       0.85      0.87      0.86        60
Sequential Art       0.98      0.82      0.89        60

      accuracy                           0.85       300
     macro avg       0.86      0.85      0.85       300
  weighted avg       0.86      0.85      0.85       300

Сохранено: confusion_matrix_bert-base-uncased.png

ИТОГОВОЕ СРАВНЕНИЕ
                  model  accuracy  f1_macro  train_time_min  epochs  batch_size
distilbert-base-uncased     0.863     0.863             3.5       3           8
      bert-base-uncased     0.847     0.849             6.6       3           8

Сохранено: model_comparison.csv


Проведено сравнение трёх подходов к классификации жанра книги по аннотации на датасете Goodreads (5 жанров, 3000 книг). Baseline на TF-IDF и логистической регрессии показал accuracy 73.3%. Дообучение DistilBERT дало 82.3% при времени обучения 3.4 минуты. BERT-base достиг наилучшего результата — 84.7% (F1 macro 0.849), но обучался почти вдвое дольше (6.6 мин). Таким образом, увеличение размера модели даёт прирост качества около 2.4 п.п., но удваивает вычислительные затраты. Для прикладной задачи с ограниченными ресурсами предпочтительнее DistilBERT; при приоритете максимальной точности — BERT-base.